### PR TITLE
Web API rewrite in Typescript, with documentation, and renames

### PR DESCRIPTION
**Pull request by Dinnerbone**
This PR covers quite a few things:

- We now generate and publish documentation with the nightlies, over to https://ruffle.rs/js-docs/master/index.html
- All JS API has been renamed from `snake_case` to `camelCase`. This is a **major breaking change**. All existing uses of the JS API will break, and negotiation between existing instances and future instances will fail. I think this is absolutely fine though, it had to be done at some point. It takes 30 seconds to just go change `stream_swf_url` to `streamSwfUrl`.
- The actual JS API is now written in Typescript and not JavaScript
- The `core` package is now aimed at publishing to npm and for websites that bundle their stuff themselves to use. For example, the demo package does that. Selfhosted is now just "core but pre-bundled in a zip form" for sites that don't have this newfangled technomancy.
- Some parts of the API have been stripped down if they aren't being used or even packaged. Let's start with a smaller API for our first release, easier to add than to remove.
- All public facing API has documentation and this is required going forwards.
- Also includes a rewritten version of #1161 as it was easier to redo it on Typescript than to redo Typescript on that

### CODE REVIEW

File: web/package.json [hunk 1]
```diff
@@ -5,31 +5,40 @@
     "license": "(MIT OR Apache-2.0)",
     "private": true,
     "devDependencies": {
+        "@typescript-eslint/eslint-plugin": "^4.7.0",
+        "@typescript-eslint/parser": "^4.7.0",
         "@wasm-tool/wasm-pack-plugin": "^1.3.1",
```
Comment (adrian17) [inline web/package.json:10]: not needed anymore?

Comment (Dinnerbone) [reply web/package.json:10]: You're right! <3

**Comment by Toad06** [comment]
I've tried your PR and everything works perfectly. 👍 I've still encountered #1517 though, I don't know if this PR could be an opportunity to fix it (it's already huge 😛).

About the autoplay PR:
- The default value for `autoplay` seems to be "off" now (it was "auto" in the previous PR). I suppose this is intended.
- The overlay is not displayed when `autoplay` is "off" and `AudioContext.state` is suspended: the "unmute_overlay" div is displayed as a block (so you can still click it and it will resume audio) but its children have a hidden visibility. This behavior is supposed to only happen when `autoplay` is "off", `unmuteOverlay` "off" as well, and `AudioContext.state` is suspended (but I was wondering if that was actually a good idea and if it wouldn't be better to provide an `audio_resume` method or something similar instead).

I've attached a demo to test autoplay: [autoplay-demo.zip](https://github.com/ruffle-rs/ruffle/files/5574262/autoplay-demo.zip)
Usage: `demo-autoplay.html?method={polyfill|whatever}&autoplay={on|off|auto}&unmuteOverlay={on|off}`

**Comment by Dinnerbone** [comment]
> I've still encountered #1517 though, I don't know if this PR could be an opportunity to fix it (it's already huge 😛).

I'll fix that!

> The default value for `autoplay` seems to be "off" now (it was "auto" in the previous PR). I suppose this is intended.

Yeah I intended it off and it wasn't super clear that it was intentionally auto before, sorry. I'm not sure which one we should go for, but I personally feel that off is better. @Herschel would you have an opinion here?


> The overlay is not displayed when `autoplay` is "off" and `AudioContext.state` is suspended: the "unmute_overlay" div is displayed as a block (so you can still click it and it will resume audio) but its children have a hidden visibility. This behavior is supposed to only happen when `autoplay` is "off", `unmuteOverlay` "off" as well, and `AudioContext.state` is suspended (but I was wondering if that was actually a good idea and if it wouldn't be better to provide an `audio_resume` method or something similar instead).

Hmm I'll look into it, I thought I copied behaviours exactly (with the exception of a bug fix when going from 'hidden' to 'visible' overlay)

**Comment by Toad06** [comment]
> I'll fix that!

Great, thanks!



> Yeah I intended it off and it wasn't super clear that it was intentionally auto before, sorry. I'm not sure which one we should go for, but I personally feel that off is better. Herschel would you have an opinion here?

No worries! Mike actually suggested it here: https://github.com/ruffle-rs/ruffle/pull/1161#issuecomment-694574399
I've personally no strong opinion about this (`off` makes sense to me currently, but as Ruffle will improve over time, `auto` may become a better option).

> Hmm I'll look into it, I thought I copied behaviours exactly (with the exception of a bug fix when going from 'hidden' to 'visible' overlay)

Yes, I don't know what's going on, everything looks good.

**Comment by Toad06** [comment]
Ah wait, the values for `unmuteOverlay` are now `visible` and `hidden`, not `on` and `off`. I've missed that, sorry, so I was actually assigning wrong values to `unmuteOverlay`. 😅 The only small issue is that the overlay visibility seems to default to `hidden` in this situation.

**Comment by Dinnerbone** [comment]
I think I've fixed the build issue, and I've changed default to auto autoplay.

**Comment by Toad06** [comment]
I confirm it works. 👍 The only remaining little thing I see is that it doesn't seem to fallback to defaults when you set a wrong value (eg. autoplay="letsgo" acts as "off").

**Comment by Dinnerbone** [comment]
Can't reproduced, tried with demo and a webserver with selfhosted ruffle.js, my extension works fine :(

**Comment by Toad06** [comment]
I've no clue what's going on too. I'll try to give you a way to reproduce but that looks like witchcraft.

Here's what I've done in Chrome (private tab) using these files ([demo-updated.zip](https://github.com/ruffle-rs/ruffle/files/5576372/demo-updated.zip)) and the compiled extension: 

- Run `demo-autoplay.html?method=polyfill` (use the polyfill): Ruffle works.
- Press F5: Ruffle crashes.
- Open the browser's console, "Network" tab, and tick the "Disable cache" box. Don't close the console. Press F5: Ruffle works (it will work every time you press F5).
- Don't close the console. Run `demo-autoplay.html` (use the JS API): Ruffle crashes.
- Untick the "Disable cache" box. Press F5: Ruffle works (it will work every time you press F5).

Yeah, I know that doesn't make any sense but I've repeated this pattern about 10 times and it produced the same results on each try.

**Comment by Dinnerbone** [comment]
Reproduced it, thank you! And I love the demo.

Looks like it's trying to load a selfhosted file from the extension, but ofc that won't work.

File: web/packages/core/src/plugin-polyfill.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+ * We also expose a method called `install` which adds a new plugin. This is
+ * used to falsify Flash detection. If the existing `navigator.mimeTypes` has an
+ * `install` method, you should not use `RuffleMimeTypeArray` as some other
+ * plugin emulator is already present.
+ */
+class RuffleMimeTypeArray implements MimeTypeArray {
+    private readonly __mimetypes: MimeType[];
+    private readonly __named_mimetypes: Record<string, MimeType>;
+
+    constructor(native_mimetype_array: MimeTypeArray | null) {
```
Comment (Herschel) [inline web/packages/core/src/plugin-polyfill.ts:17]: `native_mimetype_array` => `nativeMimeTypeArray` ?

**Comment by Dinnerbone** [comment]
I'll have to investigate this tomorrow. From what I can tell, I think this was always a problem whenever local would be considered newer than the extension, but for some reason now local is being preferred and extension was earlier.

File: web/packages/core/src/ruffle-player.ts [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+        }
+    }
+
+    private audioState(): string {
+        if (this.instance) {
+            const audioContext = this.instance.audio_context();
+            return (audioContext && audioContext.state) || "running";
+        }
+        return "suspended";
+    }
```
Comment (Herschel) [inline web/packages/core/src/ruffle-player.ts:533]: This could be a getter

File: web/packages/core/src/plugin-polyfill.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ * We can still wrap `namedItem` perfectly at least.
+ *
+ * We also expose a method called `install` which adds a new plugin. This is
+ * used to falsify Flash detection. If the existing `navigator.plugins` has an
+ * `install` method, you should not use `RufflePluginArray` as some other plugin
+ * emulator is already present.
+ */
+class RufflePluginArray {
+    private readonly __plugins: Plugin[];
+    private readonly __named_plugins: Record<string, Plugin>;
```
Comment (Herschel) [inline web/packages/core/src/plugin-polyfill.ts:111]: `__namedPlugins` ?

**Comment by Herschel** [comment]
Looks great to me (with my limited TypeScript knowledge), thank you! Your demo site is very helpful for testing @Toad06.

API thoughts:
 * Autoplay: I don't have a particularly strong opinion whether this should be "off" or "auto" by default. I guessed "auto" might be the most desired behavior, but maybe "off" is safer + less of a change for now?
 * Should `config` be passed in as a parameter of `RufflePlayer.new`?
 * What do we guarantee about changes to `config` while the player is running? Do they take effect immediately? Do we only allow them to be set once before an SWF is loaded? Maybe this varies per setting? Might be something we need to document.
 * Similarly, I am feeling more that flashVars should be included as part of the `config` Naturally they can't be changed when Flash is loaded, but I think this is fine as long as it's documented, or maybe marked readonly in the interface?
Open to opinions on any of the above.

I'd be happy to merge this when you feel it's ready, thanks for your work on this.

**Comment by Herschel** [comment]
Bikeshedding: should we rename `streamSwfUrl` and `playSwfData`? Both seem a little wordy. Since `streamSwfUrl` is what most people will want to use, I think renaming this to `load` or `loadSwf` might be good. `playSwfData` should also have a matching name, `loadData` or `loadSwfData`.

**Comment by Dinnerbone** [comment]
I've actually been considering just a generic `load({options})` method, which in javascript world makes it easier for us to keep forwards/backwards compatibility.

`player.load("url")` simplest form, shortcut for `player.load({url: "url"})`
Options could be url, data, HTTPS upgrading, scripts allowed, flashvars, etc. All of the options that don't make sense to change on the fly, but only at time of load. Then we can easily add new options without a dramatic change of API.

**Comment by Herschel** [comment]
I like the idea of `load` and allowing all of the config to be set via the `load` method, seems like that would have the best ease-of-use.

**Comment by Dinnerbone** [comment]
Let's hold that off for a future PR if you don't mind? Once we have this in place, reviewing further changes to the API will be significantly easier.

**Comment by Dinnerbone** [comment]
I've made it not crash with extension & local once again - however, I fear this issue still exists and will pop up again in the future. I've just undone the change that made it more prominent here.

**Comment by Dinnerbone** [comment]
I think this is ready to be pulled in, and we can do further work in separate PRs.

**Review by Herschel — APPROVED** [review]

**Comment by Herschel** [comment]
Let's go!

**Pull request merged**